tools/tokenize/src/main/scala/org/allenai/nlpstack/tokenize/Ai2Tokenizer.scala

```diff
@@ -0,0 +1,217 @@
+package org.allenai.nlpstack.tokenize
+
+import org.allenai.nlpstack.core.{ Token, Tokenizer }
+import org.apache.commons.lang.StringUtils
+
+import scala.collection.mutable
+
+object Ai2Tokenizer extends Tokenizer {
+  var averageTokenLength = 6 // low estimates are better
+
+  private object CharacterClass extends Enumeration {
+    type CharacterClass = Value
+    val Whitespace, Punctuation, Dash, Number, Other = Value
+  }
+  import CharacterClass._
+  private val characterClassMap = Map(
+    ' ' -> Whitespace,
+    '\n' -> Whitespace,
+    '\r' -> Whitespace,
+    '\t' -> Whitespace,
+    '\u2028' -> Whitespace,
+    '\u00A0' -> Whitespace,
+    '.' -> Punctuation,
+    ';' -> Punctuation,
+    ':' -> Punctuation,
+    '+' -> Punctuation,
+    '{' -> Punctuation,
+    '}' -> Punctuation,
+    '[' -> Punctuation,
+    ']' -> Punctuation,
+    '(' -> Punctuation,
+    ')' -> Punctuation,
+    '*' -> Punctuation,
+    '^' -> Punctuation,
+    '%' -> Punctuation,
+    '$' -> Punctuation,
+    '#' -> Punctuation,
+    '@' -> Punctuation,
+    '!' -> Punctuation,
+    '~' -> Punctuation,
+    '`' -> Punctuation,
+    ',' -> Punctuation,
+    '\'' -> Punctuation,
+    '"' -> Punctuation,
+    '—' -> Punctuation, // long dash
+    '–' -> Punctuation, // some other dash
+    '’' -> Punctuation,
+    '…' -> Punctuation,
+    '“' -> Punctuation,
+    '”' -> Punctuation,
+    '/' -> Punctuation,
+    '-' -> Dash, // normal dash
```

Comment:
Punctuation is kind of a special case because every punctuation mark becomes its own tokens. Two dashes in a row are supposed to be one token, so it needs to be its own category.